README.md

```diff
@@ -1,367 +1,654 @@
-Your installation lacks some certificates.
+## WebPush Service and States Reports
 
-1. Download [cacert.pem](http://curl.haxx.se/ca/cacert.pem).
-2. Edit your `php.ini`: after `[curl]`, type `curl.cainfo = /path/to/cacert.pem`.
+As you can see in the first example, a StatusReport object is returned.
+It can be of two types:
 
-You can also force using a client without peer verification.
+* `Minishlink\WebPush\StatusReportSuccess`: the notification has correctly been sent to the subscriber.
+* `Minishlink\WebPush\StatusReportFailure`: an error occurred.
 
-### How to solve "Bad key encryption key length" or "Unsupported key type"?
-Disable `mbstring.func_overload` in your `php.ini`.
 
-### How to solve "Class 'Minishlink\WebPush\WebPush' not found"
-Make sure to require Composer's [autoloader](https://getcomposer.org/doc/01-basic-usage.md#autoloading).
+In some cases, it could be interesting to dispatch the status report through an event dispatcher.
+The `WebPush` class has a convenient method to dispatch reports using a PSR-14 (Event Dispatcher) implementation. 
 
 ```php
-require __DIR__ . '/path/to/vendor/autoload.php';
+<?php
+use Minishlink\WebPush\ExtensionManager;
+use Minishlink\WebPush\Subscription;
+use Minishlink\WebPush\Notification;
+use Minishlink\WebPush\WebPush;
+use Psr\EventDispatcher\EventDispatcherInterface;
+
+//PSR-18 HTTP client
+/** @var Psr\Http\Client\ClientInterface $client */
+$client = '…';
+
+//PSR-17 request factory
+/** @var Psr\Http\Message\RequestFactoryInterface $requestFactory */
+$requestFactory = '…';
+
+//PSR-14 event dispatcher
+/** @var EventDispatcherInterface $eventDispatcher */
+$eventDispatcher = '…';
+
+// Extension Manager
+$extensionManager = ExtensionManager::create();
+
+$subscription = Subscription::createFromString('{"endpoint":"…"}');
+$notification = Notification::create();
+
+WebPush::create($client, $requestFactory, $extensionManager)
+    ->setEventDispatcher($eventDispatcher)
+    ->send($notification, $subscription)
+;
+
+// The status report is now dispatched (it is still returned if you need it).
 ```
 
-### I must use PHP 5.4 or 5.5. What can I do?
-You won't be able to send any payload, so you'll only be able to use `sendOneNotification($subscription)` or `queueNotification($subscription)`.
-Install the library with `composer` using `--ignore-platform-reqs`.
-The workaround for getting the payload is to fetch it in the service worker ([example](https://github.com/Minishlink/physbook/blob/2ed8b9a8a217446c9747e9191a50d6312651125d/web/service-worker.js#L75)). 
+## Complete Example
 
-### I lost my VAPID keys!
-See [issue #58](https://github.com/web-push-libs/web-push-php/issues/58).
 
-### I'm using Firebase push notifications, how do I use this library?
-This library is not designed for Firebase push notifications.
-You can still use it for your web projects (for standard WebPush notifications), but you should forget any link to Firebase while using the library.
+```php
+<?php
```

Comment:
It looks very complicated to have to set a cache and logger for each class
Couldn't we set it up once for WebPush and the library handles the rest?

Reply:
See answer below